Added ability to resize application window from all edges and corners

VRspace4 opened the pull request:
Let me know if you'd like me to make any changes.

andersevenrud commented:
Nice :)

Things like this can certainly be improved to avoid duplication:

```
  } else if (targetClassList.contains('osjs-window-resize-ne')) {
    width = getDimension(diffX, min.width, max.width, startDimension.width);
    height = getDimension(-diffY, min.height, max.height, startDimension.height);
    top = getPosition(diffY, startPosition.top, min.height, max.height, startDimension.height);
    left = startPosition.left;
   } else if (targetClassList.contains('osjs-window-resize-nw')) {
    height = getDimension(-diffY, min.height, max.height, startDimension.height);
    width = getDimension(-diffX, min.width, max.width, startDimension.width);
    left = getPosition(diffX, startPosition.left, min.width, max.width, startDimension.width);
    top = getPosition(diffY, startPosition.top, min.height, max.height, startDimension.height);
```

Very basic example:

```
const direction = target.getAttribute('data-direction');
let {width, height} = startDimension;
let {top, left} = startPosition;

if (direction.substr(0, 1) === 'n') {
    height = getDimension(-diffY, min.height, max.height, startDimension.height);
} else if (direction.substr(0, 1) === 's') {
    height = getDimension(diffY, min.height, max.height, startDimension.height);
}

if (direction.substr(-1) === 'w') {
    left = getPosition(diffX, startPosition.left, min.width, max.width, startDimension.width);
} else if (direction.substr(-1) === 'e') {
    left = startPosition.left;
}

// ...
resizeWindow({left, top, width, height});
```

andersevenrud commented:
The code quality check also failed. You might wanna look into hooking up eslint to your editor so you don't have to fix things after you've pushed :)

VRspace4 commented:
@andersevenrud I've removed the duplication the way you wanted it and fixed the linting issues, but I have to warn you this method makes it really difficult to understand the code.

andersevenrud commented on src/window-behavior.js:
```diff
@@ -255,12 +290,14 @@ export default class WindowBehavior {
         win._setState('resizing', true, false);
-        win.setDimension(getNewDimensions(
+        win.resizeWindow(getNewPositionsAndDimensions(
+          target.getAttribute('class').substring(19),
```
This feels a bit strange. If the idea is to get the last part of the string of the className you might be better of using a custom data- attribute and just straight up get the value that way. This will break if the class attribute is modified.

andersevenrud commented:
>  but I have to warn you this method makes it really difficult to understand the code.

Yeah, I'll have a glance a bit later when I have some more time. We'll work something out to make it as neat as possible :)

andersevenrud commented:
Using your changes as a starting point, I came up with this:

https://github.com/os-js/osjs-client/commit/7288038442fcbfbd59a0bf82fbaabe18b572290b

Direct link to window behaviour file, where all relevant code is: https://github.com/os-js/osjs-client/blob/7288038442fcbfbd59a0bf82fbaabe18b572290b/src/window-behavior.js

What do you think about this solution ? I think it works pretty good, and does not seem that hard to read... but I need another POV on this.

andersevenrud commented:
I also did some cleanups: https://github.com/os-js/osjs-client/commit/738e569f3f759fdc07a2f0fe98aa1be4a6e80cf0 :)

andersevenrud commented:
Also added so you cannot resize beyond the viewport https://github.com/os-js/osjs-client/commit/17ac492a8ddba2f3270e7898c2e5c5f817c127e1

andersevenrud commented:
I did another cleanup pass and did some minor adjustments. Seems to work nicely and should be easy to maintain and understand -- quite happy with the results.

> I need another POV on this.

So if you also agree, I could fork this PR and merge my code into it, so you get tagged as the original creator in the git history. Or, if I'm wrong and it's all crap, let me know :blush:

andersevenrud commented:
Since a week went by I merged this with my changes. This thread is still open for discussion though :blush:
